File: frontend/src/lib/distanceEngine.ts
/**
 * Geospatial Haversine Proximity & Transit Distance Engine (Frontend)
 */

const EARTH_RADIUS_KM = 6371;

const toRad = (deg: number) => (deg * Math.PI) / 180;

/**
 * Great-Circle Haversine distance in Kilometers
 */
export function calculateHaversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  if (lat1 === 0 && lon1 === 0) return 0;
  if (lat2 === 0 && lon2 === 0) return 0;
  if (lat1 === lat2 && lon1 === lon2) return 0;

  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) * Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const dist = EARTH_RADIUS_KM * c;

  return Number(dist.toFixed(1));
}

/**
 * Estimates emergency driving travel time in minutes with urban friction
 */
export function calculateTravelTimeMinutes(
  distanceKm: number,
  mode: "emergency" | "standard" | "transit" = "emergency"
): number {
  const d = Math.max(0, distanceKm);
  if (d === 0) return 1;

  let speedKmh = 50; // emergency driving
  let baseMinutes = 2;

  if (mode === "standard") {
    speedKmh = 35;
    baseMinutes = 4;
  } else if (mode === "transit") {
    speedKmh = 22;
    baseMinutes = 8;
  }

  const travelMins = (d / speedKmh) * 60 + baseMinutes;
  return Math.max(1, Math.round(travelMins));
}

export function formatDistance(distanceKm: number | null | undefined): string {
  if (distanceKm === null || distanceKm === undefined || isNaN(distanceKm)) {
    return "Location uncalibrated";
  }
  if (distanceKm < 1) {
    return `${Math.round(distanceKm * 1000)} m away`;
  }
  return `${distanceKm.toFixed(1)} km away`;
}

export function formatTravelTime(minutes: number | null | undefined): string {
  if (minutes === null || minutes === undefined || isNaN(minutes)) {
    return "~15 mins";
  }
  if (minutes < 60) {
    return `~${minutes} mins`;
  }
  const hrs = Math.floor(minutes / 60);
  const rem = minutes % 60;
  return `~${hrs}h ${rem}m`;
}

File: mobile/lib/core/distance_engine.dart
import 'dart:math' as math;

class DistanceEngine {
  static const double earthRadiusKm = 6371.0;

  static double calculateHaversineDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    if (lat1 == 0 && lon1 == 0) return 0.0;
    if (lat2 == 0 && lon2 == 0) return 0.0;

    final dLat = _degToRad(lat2 - lat1);
    final dLon = _degToRad(lon2 - lon1);

    final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_degToRad(lat1)) *
            math.cos(_degToRad(lat2)) *
            math.sin(dLon / 2) *
            math.sin(dLon / 2);

    final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
    final distance = earthRadiusKm * c;

    return double.parse(distance.toStringAsFixed(1));
  }

  static int calculateTravelTimeMinutes(double distanceKm, {String mode = 'emergency'}) {
    if (distanceKm <= 0) return 5;
    // Average speeds in urban emergency conditions (km/h)
    double speedKmH = 45.0;
    if (mode == 'emergency') {
      speedKmH = 60.0;
    } else if (mode == 'walking') {
      speedKmH = 5.0;
    }

    final hours = distanceKm / speedKmH;
    final mins = (hours * 60).round();
    return math.max(mins, 3); // Minimum 3 minutes
  }

  static String formatDistance(double? distanceKm) {
    if (distanceKm == null || distanceKm == 0) return 'Nearby';
    if (distanceKm < 1.0) {
      return '${(distanceKm * 1000).round()} m';
    }
    return '$distanceKm km';
  }

  static String formatTravelTime(int? minutes) {
    if (minutes == null || minutes <= 0) return '~10 mins';
    if (minutes < 60) {
      return '~$minutes mins';
    }
    final hours = minutes ~/ 60;
    final remainingMins = minutes % 60;
    return remainingMins > 0 ? '~$hours hr $remainingMins min' : '~$hours hr';
  }

  static double _degToRad(double deg) {
    return deg * (math.pi / 180.0);
  }
}
